File: Leetcode/Heap/MergeKSortedLists/MergeKSortedLists.java
package Heap.MergeKSortedLists;
// 23. Merge k Sorted Lists
// https://leetcode.com/problems/merge-k-sorted-lists/

import java.util.ArrayList;
import java.util.Arrays;
import java.util.Comparator;
import java.util.List;
import java.util.PriorityQueue;
// Definition for singly-linked list.
import LinkedList.ListNode;


// 리스트 저장 후 정렬, 9ms
class Solution {
    public ListNode mergeKLists(ListNode[] lists) {
        List<ListNode> tempList = new ArrayList<>();

        for (ListNode head : lists) {
            while (head != null) {
                tempList.add(head);
                head = head.next;
            }
        }

        tempList.sort(Comparator.comparing(node -> node.val));
        // tempList.sort((a, b) -> ((Integer) a.val).compareTo((Integer) b.val));
        // tempList.sort((a, b) -> a.val - b.val);

        for (int i = 0; i < tempList.size() - 1; ++i) {
            tempList.get(i).next = tempList.get(i + 1);
        }

        if (tempList.size() == 0) {
            return null;
        }

        tempList.get(tempList.size() - 1).next = null;
        return tempList.get(0);
    }
}


// minHeap 저장. 10ms
class Solution1 {
    public ListNode mergeKLists(ListNode[] lists) {
        PriorityQueue<ListNode> minHeap =
                new PriorityQueue<>(Comparator.comparing(node -> node.val));

        for (ListNode head : lists) {
            while (head != null) {
                minHeap.add(head);
                head = head.next;
            }
        }

        ListNode head = minHeap.poll();
        ListNode cur = head;
        while (!minHeap.isEmpty()) {
            cur.next = minHeap.poll();
            cur = cur.next;
        }

        if (cur != null) {
            cur.next = null;
        }

        return head;
    }
}


public class MergeKSortedLists {
    public static void main(String[] args) {
        testSol(new ListNode[] {new ListNode(new int[] {1, 4, 5}),
                new ListNode(new int[] {1, 3, 4}), new ListNode(new int[] {2, 6})},
                new int[] {1, 1, 2, 3, 4, 4, 5, 6});
        testSol(new ListNode[] {}, null);
    }

    static void testSol(ListNode[] input, int[] output) {
        Solution sol = new Solution();
        // todo : solution match
        ListNode temp = sol.mergeKLists(input);
        int[] res = temp != null ? temp.toIntArrays() : null;
        if (Arrays.equals(res, output)) {
            System.out.println("O : " + res);
        } else {
            System.out.println("x : " + res + "	expect : " + output);
        }
    }
}

File: Leetcode/Heap/MergeKSortedLists/MergeKSortedLists.py
# 23. Merge k Sorted Lists
# https://leetcode.com/problems/merge-k-sorted-lists/

import heapq
from typing import List, Optional


# Definition for singly-linked list.
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next


"""
자료구조를 사용하는 풀이
공간 O(1) 해법은 linkedList폴더에 있음.
"""

# O(m*n) list 104 ms. heapq 더럽다.
# 자바와 달리 다른 모듈을 불러오는게 느낌이 좋지 않음.
class Solution:
    def mergeKLists(self, lists: List[Optional[ListNode]]) -> Optional[ListNode]:
        nodeList = []

        count = 0
        for i, node in enumerate(lists):
            while node:
                count += 1
                # node.val을 비교하고 count를 비교하고 node를 비교해서 heap에 넣음.
                # 유일한 값이 아니면 node끼리 비교하려고하는데 이에 실패해서 에러남.
                heapq.heappush(nodeList, (node.val, count, node))
                node = node.next

        head = prev = ListNode()
        while nodeList:
            prev.next = heapq.heappop(nodeList)[2]
            prev = prev.next

        return head.next


# O(m*n) list 84 ms. 이게 이렇게 빨라?
class Solution:
    def mergeKLists(self, lists: List[Optional[ListNode]]) -> Optional[ListNode]:
        nodeList = []

        for node in lists:
            while node:
                nodeList.append(node)
                node = node.next

        nodeList.sort(key=lambda x: x.val)

        head = cur = ListNode()
        for node in nodeList:
            cur.next = node
            cur = cur.next
        cur.next = None
        return head.next
